File: apps/law-agent-web/lib/events/run-events.ts
import { z } from "zod";

export const runEventSchema = z.object({
  schema_version: z.literal("1.0"),
  event_id: z.string().min(1),
  event_type: z.enum([
    "run.started",
    "stage.changed",
    "token.delta",
    "citation.added",
    "approval.required",
    "run.completed",
    "run.failed",
    "stream.snapshot",
    "stream.reset",
  ]),
  run_id: z.string().min(1),
  sequence: z.number().int().nonnegative(),
  data: z.record(z.string(), z.unknown()),
});

export type RunEvent = z.infer<typeof runEventSchema>;

export function parseRunEvent(input: unknown): RunEvent {
  return runEventSchema.parse(input);
}

File: apps/law-agent-backend/src/agent/shared/contracts/run_events.py
"""Versioned contract for browser-safe run events."""

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

RunEventType = Literal[
    "run.started",
    "stage.changed",
    "token.delta",
    "citation.added",
    "approval.required",
    "run.completed",
    "run.failed",
    "stream.snapshot",
    "stream.reset",
]


class RunEvent(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["1.0"] = "1.0"
    event_id: str
    event_type: RunEventType
    run_id: str
    sequence: int = Field(ge=0)
    data: dict[str, Any]
